### DISCUSSION TITLE
Feature request: Allow Concurrent Partial Reloads with "only"

### DISCUSSION THREAD

**Discussion by rickbairstow** [inertiajs/inertia#1686] [Ideas]
Hey 👋 

We're using InertiaJS with Vue3 for our admin panel, which has several pages containing hefty amounts of data and graphs. We're diving into partial reloads to implement lazy loading when users scroll near relevant content. However, we've noticed a limitation: its not possible to do concurrent partial reloads using the "only" flag.

Inertia's feature of canceling previous requests is great and very intentional, it's been super helpful for most of our use cases so far. But isn't ideal for our larger pages where we require multiple sections' data to be lazy loaded using partial reloads. So far we've been bundling them together with `only: ['a', 'b', 'c']` but its not ideal for lazy content at that magnitude.

We're working on a feature that uses a scroll observer to trigger partial reloads for lazy data. For simplicity, let's look at this from an Vue onMounted example:

```
import { router } from '@inertiajs/vue3'

onMounted = () => {
  router.reload({ only: ['chartA'] })
  router.reload({ only: ['chartB'] })
  router.reload({ only: ['chartC'] })
}
```

We'd like to be able to run these requests simultaneously without canceling each other out, specifically when using "only", and ensure we can access all the resulting data in our Vue components.

We're interested in InertiaJS supporting concurrent partial reloads with the "only" flag. This enhancement could offer greater control over lazy loading content, especially beneficial for larger applications. Additionally, it would streamline the creation of position observers to trigger lazy loading, making it more practical and accessible for everyone.

Thanks,
Rick

**Comment by actengage** [comment]
Thank you very much for bringing up this issue. I want to post here and hopefully bring some attention to this since there are no comments. This doesn't even feel like a feature request, but a bug. It feels fundamentally broken for Inertia to do this. It requires us to use JSON/api endpoints to load things for metrics and data intensive requests. Inertia is pitched as a way to "build an app without an API" but this is pretty good example of how that's not even possible. 

For example I have a details page that loads a model, but if that model has 2 relationships I want to load into their own data tables with search fields, pagination, etc. I actually used this post to figure out why one would load, but the first request was cancelled. I don't even remotely understand how this is good behavior. In a standard SPA I can load everything concurrently. 

In summary, lazy loading should be just that and not force me to load multiple lazy loaded collections at the same time. Furthermore, this limitation isn't even mentioned in the documentation. And if it is, forgive me for missing it in the docs. This is a pretty big deal breaker for us that I feel should be getting much more attention.

**Reply by rickbairstow** [reply]
This will become a deal breaker for us soon too. We'll be working on projects that are more user-facing and might have a hefty amount of content (client controlled), lazily loading it all in one won't be feasible for these cases.

**Reply by actengage** [reply]
There are quite a few issues I have, but all are within the realm of being solvable. But it seems, at least based on reading many of the open PR's and other issues, there is no interest in doing so. It's at the point I am contemplating just forking and fixing the Javascript and maintaining my own implementation and publishing a backwards compatible alternative to Inertia if this product isn't more actively developed. 

At Laracon last year, this was billed as a "first party" package but with outside maintainers. But Taylor promised if this project goes unmaintained, they (the Laravel core team) would pick up the slack and maintain it. Taylor said "I think Laravel and Inertia is the best way to build Laravel applications". But these core problems impact any app that works at scale. We have to asynchronously load everything because we load so much data from db's with hundreds of millions of rows. We launched a version of our app without lazy loading just passing props down, and it the worst performing app I've ever built, even with very strict pagination. But lazy loading seems to be half baked to a large degree.

Another issue I have is with the entire routing system in general. Each time router.visit() is called, it creates an entirely new Axios instance (line 342 in core/router.ts). This not only creates unnecessary memory usage, but more importantly, the success/error callback doesn't even expose the Axios response.

For example, if I want to pass a header in my response (to toast a success/failure message for example), in a standard SPA I just use an Axios interceptor and use the response headers at my will. But everything must be passed down as a prop in Inertia, even though with a tiny change the axios response could be passed as a second argument to the success/error event. Nevermind that the entire Axios instance should be exposed so we can just use the instance however we want. Furthermore, i have to include another "only" prop just to get what should be available as a header. So the list of things I am required to put in "only" just keeps growing, which seems entirely antithetical to the meaning of "only".

The thing I love most about Laravel is it allows a dev to override any part of the app with the container and the design patterns used assume that everything can be extended or overridden to fix whatever use case we have. Inertia however seems to not follow these principles.

If the maintainers read this, feel free to reach out (sincerely). I have a laundry list of improvements we could make, from logic to typescript improvements. I'd even donate my time to fix these issues because we have 3 flagship products (here at ActiveEngagement) and its too late to reverse couse. I have 20 years of web development experience and used Laravel since pre 4.0. But given there are 43 open PR's and it seems to be the consensus of this product is "feature complete", I am not going to waste my time on something that isn't going to get approved.

**Reply by rickbairstow** [reply]
I forked it to play around a little and managed to get concurrent requests working easily enough. But it seems Inertia takes a snapshot of all the props when the request fires, does what it needs to, then fully replaces them. So multiple lazy requests would cause chaos. I didn't delve beyond this as I figured it'd be a domino effect of implementing one thing and potentially breaking more at Inertia's core.

**Reply by actengage** [reply]
Thanks for confirming that. That's about what I gathered too. We've gone back and forth with API and Inertia responses. It's pretty frustrating that Inertia partial reloads won't scale. I really hate to maintain API routes for certain things and break the consistency of the entire for seems like should be possible by the core library. I wish we could get some input from the maintainers on this.

**Reply by adamalbrecht** [reply]
We are also running into this limitation. @rickbairstow @actengage have you come up with any better solutions?

**Reply by rickbairstow** [reply]
I don't have any solutions for this. My use case is an internal system so we chose to accept the limitation, but our future projects will require better lazy loading and will either end up with us not using Inertia, or building our own lazy data loader of some kind.

**Reply by actengage** [reply]
I am in the same boat as @rickbairstow. We have chose to accept this limitation and shipped this into production. That said, I dont see real measurable performance hits as a result. Some pages have 5-6 requests that get cancelled and its only the last request that delivers all the data. I have forked the repo with the intent on finding a real solution, but do to time and project constraints I haven't had time to actually invest in solving this issue.
